src/mesh.cpp

```diff
@@ -142,6 +145,66 @@ vector<double> Mesh::volumes() const
   return volumes;
 }
 
+int Mesh::volume_fractions(
+  int n_sample, int bin, gsl::span<MaterialVolume> result, uint64_t* seed) const
+{
+  vector<int32_t> materials;
+  vector<int64_t> hits;
+
+#pragma omp parallel
+  {
+    vector<int32_t> local_materials;
+    vector<int64_t> local_hits;
+    Particle p;
+
+#pragma omp for
+    for (int i = 0; i < n_sample; ++i) {
+      p.r() = this->sample_element(bin, seed);
+      p.u() = {1., 0., 0.};
+      p.n_coord() = 1;
+
+      // If this location is not in the geometry at all, move on to next block
+      if (!exhaustive_find_cell(p))
+        continue;
+
+      int i_material = p.material();
+
+      // Check if this material was previously hit and if so, increment count
+      bool already_hit = false;
+      for (int j = 0; j < local_materials.size(); j++) {
+        if (local_materials[j] == i_material) {
+          local_hits[j]++;
+          already_hit = true;
+        }
+      }
+
+      // If the material was not previously hit, append an entry to the material
+      // indices and hits lists
+      if (!already_hit) {
+        local_materials.push_back(i_material);
+        local_hits.push_back(1);
+      }
+    }
+
+    // Reduce index/hits lists from each thread into a single copy
+    reduce_indices_hits(local_materials, local_hits, materials, hits);
+  }
+
+  // Make sure span passed in is large enough
+  if (hits.size() > result.size()) {
+    return -1;
```

Comment:
To trigger the expected exception for analagous method in `openmc.lib` we'll want to return `OPENMC_E_ALLOCATE` here. To that end, a test with a contrived problem containing more than 16 materials in one of the mesh elements might be useful to exercise the code in `openmc.lib.mesh`.

Reply:
This is exactly what happens in `openmc_mesh_volume_fractions` (if this method returns -1, `openmc_mesh_volume_fractions` returns `OPENMC_E_ALLOCATE`)

Reply:
Ah right, my bad. I was interpreting the return value here as an errorcode (which it kind of is?), but this isn't the CAPI function obvi.